docs/internals.rst

```diff
@@ -0,0 +1,314 @@
+  44 bytes x N ~ 220MB
+
+File chunk cache
+  probably 80-100 bytes x N ~ 400MB
+
+Encryption
+----------
+
+AES_ is used with CTR mode of operation (so no need for padding). A 64
+bits initialization vector is used, a `HMAC-SHA256`_ is computed
+on the encrypted chunk with a random 64 bits nonce and both are stored
+in the chunk. The header of each chunk is : ``TYPE(1)`` +
+``HMAC(32)`` + ``NONCE(8)`` + ``CIPHERTEXT``. Encryption and HMAC use
+two different keys.
+
+In AES CTR mode you can think of the IV as the start value for the
+counter. The counter itself is incremented by one after each 16 byte
+block. The IV/counter is not required to be random but it must NEVER be
+reused. So to accomplish this Attic initializes the encryption counter
+to be higher than any previously used counter value before encrypting
+new data.
+
+To reduce payload size only 8 bytes of the 16 bytes nonce is saved in
+the payload, the first 8 bytes are always zeros. This does not affect
+security but limits the maximum repository capacity to only 295
+exabytes (2**64 * 16 bytes).
+
+Encryption keys are either a passphrase, passed through the
+``ATTIC_PASSPHRASE`` environment or prompted on the commandline, or
+stored in automatically generated key files.
+
+Key files
+---------
+
+When initialized with the ``init -e keyfile`` command, |project_name|
+needs an associated file in ``$HOME/.attic/keys`` to read and write
+the repository. The format is based on msgpack_, base64 encoding and
+PBKDF2_ SHA256 encryption, which is then encoded again in a msgpack_.
+
+The internal data structure is as follows:
+
+version
+  currently always an integer, 1
+
+repository_id
+  the ``id`` field in the ``config`` ``INI`` file of the repository.
+
+enc_key
+  the key used to encrypt data with AES (256 bits)
+  
+enc_hmac_key
+  the key used to HMAC the resulting AES-encrypted data (256 bits)
+
+id_key
+  the key used to HMAC the above chunks, the resulting hash is
+  stored out of band (256 bits)
+
+chunk_seed
+  the seed for the buzhash chunking table (signed 32 bit integer)
+
+Those fields are encoded using msgpack_. The utf-8-encoded phassphrase
+is encrypted with PBKDF2_ and SHA256_ using 100000 iterations and a
+random 256 bits salt to give us a derived key. The derived key is 256
+bits long.  A `HMAC-SHA256`_ checksum of the above fields is generated
+with the derived key, then the derived key is also used to encrypt the
+above pack of fields. Then the result is stored in a another msgpack_
+formatted as follows:
+
+version
+  currently always an integer, 1
+
+salt
+  random 256 bits salt used to encrypt the passphrase
+
+iterations
+  number of iterations used to encrypt the passphrase (currently 100000)
+
+algorithm
+  the hashing algorithm used to encrypt the passphrase and do the HMAC
+  checksum (currently the string ``sha256``)
+
+hash
+  the HMAC checksum of the encrypted derived key
+
+data
+  the derived key, encrypted with AES over a PBKDF2_ SHA256 hash
```

Comment:
s/hash/key/ - it is a KDF (key derivation function), so its output is a key. It uses a hash internally.